_source/_posts/2024-04-30-spring-api-fga.md

```diff
@@ -0,0 +1,901 @@
+
+Run the test with:
+
+```shell
+./gradlew test --tests com.example.demo.DocumentIntegrationTest
+```
+
+## Running the Spring Boot API
+
+Taking advantage of Spring Boot Docker, create a `compose.yml` file at the root of the project, to start an OpenFGA server when the application runs:
+
+```yaml
+# compose.yml
+services:
+  openfga:
+    image: openfga/openfga:v1.5.3
+    container_name: openfga
+    command: run
+    environment:
+      - OPENFGA_DATASTORE_ENGINE=memory
+      - OPENFGA_PLAYGROUND_ENABLED=true
+    networks:
+      - default
+    ports:
+      - "8090:8080" #http
+      - "3000:3000" #playground
+    healthcheck:
+      test: ["CMD", "/usr/local/bin/grpc_health_probe", "-addr=openfga:8081"]
+      interval: 5s
+      timeout: 30s
+      retries: 3
+```
+
+Notice the HTTP API port is mapped to host port `8090`. Run the application with:
+
+```shell
+./gradlew bootRun
+```
+
+You should see in the application logs messages when the OpenFGA container is starting:
+```
+[Document API] [           main] .s.b.d.c.l.DockerComposeLifecycleManager : Using Docker Compose file '.../compose.yml'
+[Document API] [utReader-stderr] o.s.boot.docker.compose.core.DockerCli   :  Container openfga  Recreate
+[Document API] [utReader-stderr] o.s.boot.docker.compose.core.DockerCli   :  Container openfga  Recreated
+[Document API] [utReader-stderr] o.s.boot.docker.compose.core.DockerCli   :  Container openfga  Starting
+[Document API] [utReader-stderr] o.s.boot.docker.compose.core.DockerCli   :  Container openfga  Started
+[Document API] [utReader-stderr] o.s.boot.docker.compose.core.DockerCli   :  Container openfga  Waiting
+[Document API] [utReader-stderr] o.s.boot.docker.compose.core.DockerCli   :  Container openfga  Healthy
+```
+
+Create a new test token with Auth0 CLI:
+
+```shell
+auth0 test token -a https://document-api.okta.com -s openid
+```
+
+With curl, send a request to the API server using a bearer access token:
+
+```shell
+TOM_ACCESS_TOKEN=<auth0-access-token>
+```
+
+```shell
+curl -i -X POST \
+  -H "Authorization:Bearer $TOM_ACCESS_TOKEN" \
+  -H "Content-Type: application/json" \
+  -d '{"name": "planning.doc"}' \
+  http://localhost:8080/document
+```
+
+Verify the API does not authorize creating a document with a parent not owned:
+
+```shell
+curl -i -X POST \
+  -H "Authorization:Bearer $TOM_ACCESS_TOKEN" \
+  -H "Content-Type: application/json" \
+  -d '{"name": "planning.doc", "parentId": 6}' \
+  http://localhost:8080/document
+```
+
+The API will reject with HTTP 403:
+
+```
+HTTP/1.1 403
+...
+Access is denied%                      
+```
+
+## Authorizing Access
+
+You can remove the auth0.com cookie, create an access token for a different user, and request a document not owned with the following command:
+
+```shell
+curl -i -H "Authorization:Bearer $ANA_ACCESS_TOKEN" http://localhost:8080/document/4
```

Comment:
where is `ANA_ACCESS_TOKEN` set and what doed `ANA` mean here its a bit confusing, I assume its just access token for a user named Ana ? But seeing this out of context is confusing